applications/MeshingApplication/custom_utilities/mmg_utility.h

```diff
@@ -2244,6 +2292,68 @@ class MmgUtility
     /***********************************************************************************/
     
     template<>  
+    void MmgUtility<2>::CalculateInitialCoordinates(
+        NodeType::Pointer& pNode,
+        const ElementType::Pointer& pElement,
+        const Vector shape_functions
+        )
+    {
+        double& initial_coord_X = pNode->X0();
+        double& initial_coord_Y = pNode->Y0();
+        
+        const array_1d<double, 3>& node0_initial_coord = pElement->GetGeometry()[0].Coordinates() - pElement->GetGeometry()[0].FastGetSolutionStepValue(DISPLACEMENT, 0);
```

Comment:
also a question here: if you assume, as you do, that nodes have displacement (which is something you should check at least for the first node), then you could use the interpolated value to reconstruct the original position.

i mean:
node.X0 = node.X - node.GetSolutionStepValue(DISPLACEMENT)

that would be faster, and consistent with the interpolation of displacements